Refactor PFDL code

maxhoerstr opened the pull request:
- Make code more modular to enable the reuse of the code (e.g.. make more functions and methods single purpose)
- Fix various bugs in the petri net generation
- Renaming
- Format code and add docstrings
- Allow a PFDL program to be passed over to the scheduler as a string (currently, only a file path was allowed)
- Add various quality of live features/fixes
- Allow the free configuration of the "start task" (former productionTask)
- Allow integers as loop limits and negative numbers

maxhoerstr committed:
Do not generate callbacks if used in extension

maxhoerstr committed:
Save petri net as dot file in extension

maxhoerstr committed:
Remove unnecessary new line character

maxhoerstr committed:
Add group ids for places in petri net

maxhoerstr committed:
Save group ids in transitions

maxhoerstr committed:
Add id of parent group

maxhoerstr committed:
Fix missing args in method calls

maxhoerstr committed:
Create a tree with group ids while generating

- Create a tree where each node contains a group id, to map the
call hierarchy of the different tasks

maxhoerstr committed:
Save petri net tree into media folder

maxhoerstr committed:
Remove now unused parent group ids

There were still parent group ids used.
They are unnecessary now, since the structure of the calls is realised
with a tree structure (Node)

maxhoerstr committed:
Save petri net tree in dot file

maxhoerstr committed:
Fix drawing of petri net

- Petri Net images and dot files are no always named "petri_net"
- Append Call Tree from PetriNetGenerator to dot file for extension

maxhoerstr committed:
Add identifier in dot file for call tree

maxhoerstr committed:
Remove comment in dot file

maxhoerstr committed:
Enforce passing of file name for extension

maxhoerstr committed:
Fix bugs

maxhoerstr committed:
Cluster petri net components

- Cluster individual components of the PFDL (Service, Tasks, ..)
- Save this information in the dot file

maxhoerstr committed:
Draw dot file every time

maxhoerstr committed:
Fix minor issues

- Fix bug in drawing of petri net for extension
- Move creation of folders on better suiting location
- Remove unnecessary `used_in_extension` attributes and params

maxhoerstr committed:
Minor changes

- Fix typos and remove comments
- Temporarily disable removal of nodes from the petri net
- Publish PFDL string to dashboard
- Rewrite methods so PFDL string can be extracted from the file and used

maxhoerstr committed:
Small fixes

- Send Petri Net update before Order is finished
- Remove counting loop subgraph in dot file

maxhoerstr committed:
Save order specific petri net image and dot file

maxhoerstr committed:
Use scheduler id for logging

maxhoerstr committed:
Remove saving of tokens

maxhoerstr committed:
Use correct transition list

maxhoerstr committed:
Cast loop limit to int before using it

maxhoerstr committed:
Add correct casting

maxhoerstr committed:
Allow integers in counting loop limits

maxhoerstr committed:
Support integer loop limits in scheduler

maxhoerstr committed:
Move content of init into separate functions

This allows for more modularity so future plugin users can choose what
to reuse if the overwrite the init function of the scheduler.

maxhoerstr committed:
Allow other statements to be processed

maxhoerstr committed:
Check if task and service are in loop in callbacks

- Only assign new uuids to task and service apis if they are in a loop
- Assign test ids correctly

OliverStolzBO committed:
Add option to pass the PFDL program as a string

OliverStolzBO committed:
Rename id to uuid for consistency

- Use the term "uuid" wherever a uuid is assigned
- Avoid inconsistent variable naming

OliverStolzBO committed:
Add helper method to cast datatypes from string

OliverStolzBO committed:
Enable empty comments in PFDL programs

OliverStolzBO committed:
Move helper method into semantic error checker

- Enable plugins to overload this method

maxhoerstr committed:
Add unit test for integer loop limits

OliverStolzBO committed:
Rename dict key for consistency

OliverStolzBO committed:
Edit test file to avoid naming conflicts

- the changed file might have caused troubles when running it while the
MFDL plugin is active

OliverStolzBO committed:
Split method

- Enable reusing parts of the method when multiple types per variable
 are allowed

maxhoerstr committed:
Correct the clustering of nodes

The way that the nodes (places and transitions) in the petri net
were clustered with the help of SNAKES was wrong. Now, the cluster is not overwritten anymore, but correctly added in the intended way.
This resolves a bug where the deletion of specific nodes resulted in an KeyError in the snakes library.

OliverStolzBO committed:
Update generator

- Adjust docstrings
- Remove unused attribute 'cluster'

OliverStolzBO committed:
Extend grammar by variable type representation

- Add definition for variable type
- Apply changes to tree visitor
- Adjust tests

maxhoerstr committed:
Fix old clustering code

maxhoerstr committed:
Allow negative numbers

maxhoerstr committed:
Update generated lexer file

maxhoerstr committed:
Add better documentation for petri net logic

OliverStolzBO committed:
Enable flexible starting task names for plugins

- Instead of limiting the starting task name to 'productionTask',
it can now be modified by passing the desired name to the Process

maxhoerstr committed:
Merge branch 'main' into refactoring

maxhoerstr committed:
Update transition ids that were broken after merge

maxhoerstr committed:
Remove more code that was introduced due to the merge

maxhoerstr committed:
Add constants that were lost due to merge

maxhoerstr committed:
Check if there is only a single task call in parallel loop

OliverStolzBO commented on tests/test_files/scheduling/task_synchronisation.pfdl:
```diff
@@ -20,7 +20,7 @@ Struct CuttingResult
 End
 
-Struct TransportOrderStep
+Struct TransportOrderStep_Struct
```
Maybe remove the _ in the Struct and Service names here to be consistent with the other test files

OliverStolzBO commented on tests/test_files/scheduling/task_synchronisation.pfdl:
```diff
@@ -50,14 +50,14 @@ Task cuttingTask
 
 Task transport
-    Transport
+    Transport_Service
```
See comment above

OliverStolzBO commented on tests/unit_test/test_pfdl_tree_visitor.py:
```diff
@@ -651,51 +671,77 @@ def test_visitVariable_definition(self):
+
+        # without length
+        with patch.object(self.visitor, "visitArray", return_value=-1):
+            array = Array("string")
+            array.length = 10
```
This lines are meaningless and can be removed

OliverStolzBO reviewed (commented):
Looks good overall, I only figured out some minor improvements that could be applied to increase the code quality.

maxhoerstr commented on pfdl_scheduler/petri_net/generator.py:
```diff
@@ -7,6 +7,7 @@
 
 # standard libraries
+import copy
 from typing import Any, Callable, Dict, List, OrderedDict, Tuple
```
Copy is used in Line 54 :)

OliverStolzBO replied:
Sorry I marked the wrong line, this comment refers to the Tuple import in the line below

maxhoerstr committed:
Incorporate PR feedback

Incorporate feedback regarding the quality of the code (fix typos, remove comments/ unused packages, fix naming)

OliverStolzBO commented on tests/test_files/scheduling/task_synchronisation.pfdl:
```diff
@@ -20,7 +20,7 @@ Struct CuttingResult
 End
 
-Struct TransportOrderStep_Struct
+Struct TransportOrderStep
```
Using keywords of other plugins will cause problems there, so the Struct needs to have a different name than TransportOrderStep (same for Transport)

OliverStolzBO committed:
Rename test program variables
